gc/base/EvacuatorWorklist.hpp

```diff
@@ -0,0 +1,325 @@
+		_head = free;
+		_count += 1;
+		Debug_MM_true((0 == _count) == (NULL == _head));
+	}
+
+	/**
+	 * Claim 0 or more free elements from system (forge) memory or C++ stack space
+	 *
+	 * @return the number of free elements added
+	 */
+	MMINLINE uintptr_t
+	refresh()
+	{
+		if ((NULL == _underflow) || (0 == _count)) {
+			/* allocate a new underflow chunk and link it in at head of underflow list*/
+			UnderflowChunk *underflowChunk = (UnderflowChunk *)_forge->allocate(sizeof(UnderflowChunk), OMR::GC::AllocationCategory::FIXED, OMR_GET_CALLSITE());
+			underflowChunk->next = _underflow;
+			_underflow = underflowChunk;
+
+			/* add free elements from underflow chunk to free list */
+			MM_EvacuatorWorkPacket *packet = underflowChunk->packet;
+			MM_EvacuatorWorkPacket *end = packet + work_packet_chunksize;
+			while (packet < end) {
+				add(packet);
+				packet += 1;
+			}
+		}
+
+		Debug_MM_true(0 < _count);
+		return _count;
+	}
+
+	MMINLINE void
+	reload()
+	{
+		if (NULL != _underflow) {
+			_count = 0;
+			_head = NULL;
+			for (UnderflowChunk *underflowChunk = _underflow; NULL != underflowChunk; underflowChunk = underflowChunk->next) {
+				/* add free elements from underflow chunk to free list */
+				MM_EvacuatorWorkPacket *packet = underflowChunk->packet;
+				MM_EvacuatorWorkPacket *end = packet + work_packet_chunksize;
+				while (packet < end) {
+					Debug_MM_true(NULL == packet->base);
+					add(packet);
+					packet += 1;
+				}
+			}
+		} else {
+			refresh();
+		}
+	}
+
+	/**
+	 * Releases all underflow chunks, if any are allocated, and resets list to empty state.
+	 */
+	MMINLINE void
+	reset()
+	{
+		/* deallocate all underflow chunks allocated from system memory (forged) */
+		while (NULL != _underflow) {
+			UnderflowChunk *next = _underflow->next;
+			_forge->free(_underflow);
+			_underflow = next;
+		}
+		_head = NULL;
+		_count = 0;
+	}
+
+	/**
+	 * Constructor.
+	 */
+	MM_EvacuatorFreelist(MM_Forge *forge)
+		: MM_Base()
+		, _head(NULL)
+		, _count(0)
+		, _underflow(NULL)
+		, _forge(forge)
+	{ }
+};
+
+/**
+ * FIFO queue of work packets. Other threads may access this list if their own queues run dry.  The
+ * queue volume is modified only within critical regions guarded by a mutex declared by the evacuator
+ * that owns the queue. It is exposed to other evacuator threads as a volatile value to enable them to
+ * check queue volume without taking the mutex.
+ */
+class MM_EvacuatorWorklist : public MM_Base
+{
+/*
+ * Data members
+ */
+private:
+	MM_EvacuatorWorkPacket *_head;	/* points to head of worklist */
+	MM_EvacuatorWorkPacket *_tail;	/* points to tail of worklist */
+	volatile uint64_t _volume;		/* (bytes) volume of work contained in list */
+
+protected:
+public:
+
+/*
+ * Function members
+ */
+private:
+protected:
+public:
+	/**
+	 * Returns a pointer to the volatile sum of the number of bytes contained in work packets in the list
+	 */
+	MMINLINE volatile uint64_t *volume() { return &_volume; }
```

Comment:
Is this returning the address of the value so other code can modify the contents?  If so wouldn't it make more set to add an API to update the volume?

Reply:
Worklist volume is adjusted whenever a work packet is added or removed. It is not externally settable.

Reply:
I see code using _volume directly so I am still confused why this API exists in its current form.

Reply:
_volume is a volatile private member of MM_EvacuatorWorklist and is modified directly only within that class. The volume() method is public to allow MM_Evacuator instances to assess the volume of other evacuator's worklist. It returns a pointer to this volatile member only to ensure that the caller obtains the volume value only when it dereferences the pointer. To return the value directly from the volume() method could result is use of a stale value when returned value is applied.